### PR TITLE
Add support of AMQP queues

### PR DESCRIPTION
**Pull request by mekahell**
I used the existing implementations of EventQueueProvider and ObservableQueue to add the support for AMQP queues. 
I've used the version 3.7.13 of RabbitMQ server.
Are you interested ?
Thanks for all the work you do and you done on Conductor :)

### CODE REVIEW

**Comment by jkaipa** [comment]
> I used the existing implementations of EventQueueProvider and ObservableQueue to add the support for AMQP queues.
> I've used the version 3.7.13 of RabbitMQ server.
> Are you interested ?
> Thanks for all the work you do and you done on Conductor :)

@mekahell Can you consider exchange and routing key for publishing instead of using a queue name? I would consider the format for sink as amqp:exchange_name:routing_key, but for consumption it would be a queue name in the format exchange_name:queue_name

**Comment by mekahell** [comment]
@jkaipa Your solution and mine don't let the users choose between exchanges or queues. Maybe we can use the URI syntax to specify the name of exchange and its type for publishing.

Example of URI:

amqp:myQueueName?publishTarget=exchange&publishName=myPublishExchange&publishType=topic&publishRoutingKey=myRoutingKey

**Comment by jkaipa** [comment]
Can you please point me to the specification/documentation for this format to specify queue and exchange?

**Comment by mekahell** [comment]
@jkaipa I've made some changes according to your review. 
There are 2 types of URI scheme for AMQP queues :
- **amqp-queue** for queue
- **amqp-exchange** for exchange

You can append parameters to the URI to set the following properties:

- "exchangeType" : type of exchange topic, fanout, direct,...
- "routingKey" : the routing key
- "deliveryMode" : the delivery mode of message to publish
- "durable" : is queue durable ? True by default
- "exclusive" : is queue exclusive? False by default
- "autoDelete" : is auto delete enabled ? False by default
- "maxPriority" : Enable the priority feature on queue and set the max priority

For more details about the query parameters please see the Enum class com.netflix.conductor.contribs.queue.amqp.AMQQueryParameters and the unit tests.

**Comment by jkaipa** [comment]
> @jkaipa I've made some changes according to your review.
> There are 2 types of URI scheme for AMQP queues :
> 
> * **amqp-queue** for queue
> * **amqp-exchange** for exchange
> 
> You can append parameters to the URI to set the following properties:
> 
> * "exchangeType" : type of exchange topic, fanout, direct,...
> * "routingKey" : the routing key
> * "deliveryMode" : the delivery mode of message to publish
> * "durable" : is queue durable ? True by default
> * "exclusive" : is queue exclusive? False by default
> * "autoDelete" : is auto delete enabled ? False by default
> * "maxPriority" : Enable the priority feature on queue and set the max priority
> 
> For more details about the query parameters please see the Enum class com.netflix.conductor.contribs.queue.amqp.AMQQueryParameters and the unit tests.

The next point is about the consumption. Is there a reason to go with basic get vs basic consume? How about doing a basic consume like below? (I didn't add additional params such as prefetchCount etc., for now). Conductor would be running much faster with its task execution. If the messages are getting into rabbit success or failure queue at faster rate, basicConsume might deliver better performance.

```

	@VisibleForTesting
	public void receiveMessages() {
		try {
			DeliverCallback deliverCallback = (consumerTag, delivery) -> {
				Message message = new Message();
				long deliveryTag = delivery.getEnvelope().getDeliveryTag();
				String response = new String(delivery.getBody(), "UTF-8");
				if (response != null) {
					message.setId(NUID.nextGlobal());
					message.setPayload(response);
					message.setReceipt("rec");
					message.setDeliveryTag(deliveryTag);
					messages.add(message);
					logger.debug("Message received. {}", message.getPayload());
				}
			};

			channel.basicConsume(queueName, false, deliverCallback, consumerTag -> {
			});
			Monitors.recordEventQueueMessagesProcessed(QUEUE_TYPE, this.queueName, messages.size());
		} catch (Exception e) {
			logger.error("Exception while getting messages from Rabbit ", e);
			Monitors.recordObservableQMessageReceivedErrors(QUEUE_TYPE);
		}
	}

	@VisibleForTesting
	public Observable.OnSubscribe<Message> getOnSubscribe() {
		receiveMessages();
		return subscriber -> {
			Observable<Long> interval = Observable.interval(pollTimeInMS, TimeUnit.MILLISECONDS);
			interval.flatMap((Long x) -> {
				List<Message> available = new LinkedList<>();
				messages.drainTo(available);
				if (!available.isEmpty()) {
					AtomicInteger count = new AtomicInteger(0);
					StringBuilder buffer = new StringBuilder();
					available.forEach(msg -> {
						buffer.append(msg.getId()).append("=").append(msg.getPayload());
						count.incrementAndGet();
						if (count.get() < available.size()) {
							buffer.append(",");
						}
					});

					logger.info(String.format("Batch from %s to conductor is %s", queueName, buffer.toString()));
				}
				return Observable.from(available);
			}).subscribe(subscriber::onNext, subscriber::onError);
		};
	}
```

**Comment by mekahell** [comment]
Hi @jkaipa I'm totally agree but It was easier to me to use basicGet with rx.Observable instead of basicConsume :( Execution of method basicGet is synchronous and basicConsume is async.

**Comment by jkaipa** [comment]
@mekahell Appreciate your thoughts. We ran several load tests with both and figured improved performance with basicConsume.

File: contribs/src/main/java/com/netflix/conductor/contribs/AMQModule.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    private static Logger logger = LoggerFactory.getLogger(AMQModule.class);
+
+    @Override
+    protected void configure() {
+        bind(QueueManager.class).asEagerSingleton();
+        logger.info("RabbitMQ Module configured ...");
+    }
+
+    @ProvidesIntoMap
+    @StringMapKey("amqp-queue")
```
Comment (apanicker-nflx) [inline contribs/src/main/java/com/netflix/conductor/contribs/AMQModule.java]: Please rename to `amqp` to be consistent with other queue implementations.

File: contribs/src/main/java/com/netflix/conductor/contribs/AMQModule.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    @Singleton
+    @Named(EVENT_QUEUE_PROVIDERS_QUALIFIER)
+    public EventQueueProvider getAMQQueueEventQueueProvider(Configuration config) {
+        return new AMQEventQueueProvider(config, false);
+    }
+
+    @ProvidesIntoMap
+    @StringMapKey("amqp-exchange")
```
Comment (apanicker-nflx) [inline contribs/src/main/java/com/netflix/conductor/contribs/AMQModule.java]: Please rename to `amqp_exchange`

File: versionsOfDependencies.gradle [hunk 3]
```diff
@@ -59,4 +59,5 @@ ext {
     revSlf4j = '1.7.25'
     revSlf4jlog4j = '1.8.0-alpha1'
     revSpectator = '0.68.0'
+    revAmqpClient = '5.6.0'
```
Comment (apanicker-nflx) [inline versionsOfDependencies.gradle]: Please keep this file in alphabetical order.

**Comment by apanicker-nflx** [comment]
@mekahell Thanks for the contribution. 
@jkaipa Thank you for taking the time to review the changes. Based on the discussion above, we will wait for you to sign off on the changes before merging as the changes appear to be innocuous to execution unless specifically included for use.

**Comment by tburschka** [comment]
Can't wait to see this feature

File: versionsOfDependencies.gradle (hunk 3 above)
Comment (mekahell) [reply versionsOfDependencies.gradle]: I've just pushed to modifications

**Comment by tburschka** [comment]
@apanicker-nflx any news here?

**Comment by apanicker-nflx** [comment]
@tburschka Since, we are not using RabbitMQ, we haven't been able to test this feature out. However, we would appreciate help from the community with testing this. 
Please feel free to review the PR and we will be merging this soon.
Also, paging @jkaipa based on the discussion earlier in the PR.

**Comment by jkaipa** [comment]
@mekahell @apanicker-nflx My suggestion of using basicConsume can be implemented through an enhancement. Rest is looking great.

**Comment by kishorebanala** [comment]
Hey @mekahell, we had to revert this PR as the build have been failing with couple of tests. Do you mind taking a look? https://travis-ci.org/Netflix/conductor/builds/568130006

**Pull request merged**